src/sage/interfaces/singular.py

```diff
@@ -1466,10 +1466,34 @@ def __len__(self):
         EXAMPLES::
 
             sage: R = singular.ring(0, '(x,y,z)', 'dp')
-            sage: A = singular.matrix(2,2)
+            sage: f = singular.poly('x+4*y+6'); f
+            x+4*y+6
+            sage: len(f)
+            3
+
+            sage: v = singular.vector('[x,y,4]'); v
+            [x,y,4]
+            sage: len(v)
+            3
+
+        Beware that vectors are truncated when ending with zeros::
+
+            sage: v = singular.vector('[x,y,0,0]'); v
+            [x,y]
+            sage: len(v)
+            2
+
+        For matrices, the length is the number of columns::
+
+            sage: A = singular.matrix(2, 4)
             sage: len(A)
             4
+            sage: A = singular.matrix(2, 2)
+            sage: len(A)
+            2
         """
+        if self.type() == 'matrix':
+            return int(self.ncols())
```

Comment:
This is technically a breaking change, but it seems to make sense since it's consistent with `list`. Do leave an explanation though. (on the other hand there's the question why not the other option i.e. make `list` return the flattened list)